### PR TITLE
WIP: Add KEP for PersistentVolume release timestamp

### PR DESCRIPTION
**Pull request by RomanBednar**

### CODE REVIEW

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md [hunk 1]
```diff
@@ -0,0 +1,878 @@
...
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+We want to add a new PersistentVolumeStatus field, which would hold a timestamp of when the PersistentVolume was moved
+to `Released` state.
+
+## Motivation
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Michelle wanted a generic "phase change timestamp" that could help us to know e.g. how long it took to bind a volume. That should go into motivation + the field name + the KEP title.

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md [hunk 2]
```diff
@@ -0,0 +1,878 @@
...
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+We need to update api server to support the newly proposed field and update PV controller to set a value of the new
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: the API server

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md [hunk 3]
```diff
@@ -0,0 +1,878 @@
...
+-->
+
+The new field is purely informative and should not introduce any risk.
+
+## Design Details
+
+Changes required for this KEP:
+
+* kube-apiserver
+  * extend PersistentVolumeStatus type with `ReleaseTimestamp` field
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: `LastPhaseTransitionTime`?

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md [hunk 4]
```diff
@@ -0,0 +1,878 @@
...
+
+* kube-controller-manager / PV controller
+  * update the timestamp whenever PV controller updates volume status to `VolumeReleased`
+    * Validation should check timestamp format 
+    * Validation should allow update from `nil`
+    * Validation should not allow updating to a point in time before the current timestamp
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Validation is done in the API server

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md [hunk 5]
```diff
@@ -0,0 +1,878 @@
...
+
+## Design Details
+
+Changes required for this KEP:
+
+* kube-apiserver
+  * extend PersistentVolumeStatus type with `ReleaseTimestamp` field
+
+* kube-controller-manager / PV controller
+  * update the timestamp whenever PV controller updates volume status to `VolumeReleased`
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: update the timestamp whenever PV controller changes Phase of a PV.

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md [hunk 6]
```diff
@@ -0,0 +1,878 @@
...
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+No change in cluster upgrade / downgrade process.
+
+When downgrading from a version that added the new timestamp field to persistent volumes
+we need to make sure that after downgrade the disabled fields are removed from the persistent
```
Comment (RomanBednar) [inline keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: I changed this too - we don't what to remove the field in this case but just set it's value to `nil`.

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md (hunk 5 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Done, should "phase" be capitalized? And for all mentions? I think we don't need to refer to the field itself with "Phase".

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md (hunk 3 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Sounds good, let's use it in feature gate name too so the naming is consistent.

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md (hunk 2 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Done.

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md (hunk 1 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Done. I've changed it throughout the KEP so it's more generic and kept the original use case with released PVs as an example.

File: keps/sig-storage/3762-persistent-volume-release-timestamp/README.md (hunk 4 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-release-timestamp/README.md]: Moved under `kube-apiserver`.

**Comment by RomanBednar** [comment]
@jsafrane I've updated the KEP based on your comments and used more consistent terminology:

1) use "PV" instead of "persistent volume" (or mixing both)
2)  s/change phase/transition phase
3) title "PersistentVolume last phase transition time"
4) feature gate "PersistentVolumeLastPhaseTransitionTime"

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+## Motivation
+
+Some users have experienced data loss when using `Delete` retain policy and reverted to a safer `Retain` policy.
+With `Retain` policy all volumes that are retained and left unclaimed have their phase is set to `Released`.
+As the released volumes pile up over time admins want to perform manual cleanup based on the time when the volume was
+last used, which is when the volume transitioned to `Released` phase.
+
+We can approach the solution in a more generic way and record a timestamp of when the volume transitioned to any phase,
+not just to `Released` phase. This is more informative for admins as it covers more scenarios while it still meets the
+original expectation described above.
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]:
```suggestion
not just to `Released` phase. This allows anyone, incl. our perf tests, to measure time e.g. between a PV `Pending` and `Bound`
```

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md [hunk 8]
```diff
@@ -0,0 +1,885 @@
...
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+1) Introduce a new status field in PersistentVolumes.
+2) Update the new field with a timestamp every time a volume transitions to a different phase.
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]:
```suggestion
2) Update the new field with a timestamp every time a volume transitions to a different phase (`pv.Status.Phase`).
```

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+bogged down.
+-->
+
+#### Story 1
+
+As a cluster admin I want to use `Retain` policy for released volumes which is safer than `Delete` and gives me more
+control over volume cleanup. However, this in turn leads to a growing number of volumes that need to be monitored
+and cleaned up. The cleanup should be performed based how long the volume was not used by any claim. I should be able
+to list all volumes that are now in `Released` phase and see when it transitioned to this phase to decide which volumes
+can be safely removed.
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]:
```suggestion
As a cluster admin I want to use `Retain` policy for released volumes, which is safer than `Delete`, and implement a reliable policy to delete volumes that are `Released` for more than X days.
```

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md [hunk 10]
```diff
@@ -0,0 +1,885 @@
...
+Changes required for this KEP:
+
+* kube-apiserver
+  * extend PersistentVolumeStatus type with `LastPhaseTransitionTime` field
+  * validation should check timestamp format 
+  * validation should allow update from `nil`
+  * validation should not allow updating to a point in time before the current timestamp
+
+* kube-controller-manager / PV controller
+  * update the timestamp whenever PV controller transitions PV to a different phase
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]:
```suggestion
  * update the timestamp whenever PV controller transitions PV to a different phase (`pv.Status.Phase`)
```

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md [hunk 11]
```diff
@@ -0,0 +1,885 @@
...
+  * remove the timestamp in `LastPhaseTransitionTime` field during volume status update when feature gate is disabled
+    * extend and use the existing function for disabling a status field: https://github.com/kubernetes/kubernetes/blob/d94261e904c90578e913c42c4d2a0fc8cb30937f/pkg/api/persistentvolumeclaim/util.go#L91
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]: I think util.go is used in the API server in PrepareForUpdate / Create. Do you really want the controller to clear the field proactively?

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md [hunk 12]
```diff
@@ -0,0 +1,885 @@
...
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [X] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name:
+  - Components depending on the feature gate:
```
Comment (jsafrane) [inline keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]: File this please

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md (hunk 11 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]: Removing this from KEP for now - it does not have to be that specific probably, I'll figure out the right place later.

File: keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md (hunk 12 above)
Comment (RomanBednar) [reply keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md]: Done.

**Pull request closed without merging**